csharp/src/Ice/InputStream.cs

```diff
@@ -81,8 +65,22 @@ internal int Pos
             }
         }
 
-        // Number of bytes remaining in the underlying buffer.
-        private int Remaining => _limit - Pos ?? _buf.B.Remaining();
+        /// <summary>The position (offset) in the underlying buffer.</summary>
+        internal int Pos
+        {
+            get => _tail;
+            set
+            {
+                if (value < 0 || value > _buffer.Count)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(Pos), "The position value is outside the buffer bounds.");
+                }
+                _tail = value;
+            }
+        }
+
+        /// <summary>Returns the current size of the stream.</summary>
+        internal int Size => _buffer.Count;
```

Comment:
`Count` would be a better property name, but no need to change it since we'll refactor all this soon.